src/jit/morph.cpp

```diff
@@ -7411,6 +7411,41 @@ void Compiler::fgMorphTailCallViaHelper(GenTreeCall* call, void* pfnCopyArgs)
     if (call->IsVirtualStub())
     {
         call->gtFlags |= GTF_CALL_NULLCHECK;
+
+#if defined(_TARGET_AMD64_)
+        // If we already inited arg info here then we will have added the VSD
+        // arg on AMD64. So we remove it here as we will handle this case
+        // specially below.
+        fgArgInfo* argInfo = call->fgArgInfo;
+        if (argInfo != nullptr)
+        {
+            GenTreeCall::Use* vsdArg = nullptr;
+
+            for (unsigned index = 0; index < argInfo->ArgCount(); ++index)
+            {
+                fgArgTabEntry* arg = argInfo->GetArgEntry(index, false);
+                if (arg->isNonStandard)
+                {
+                    // The only supported nonstandard arg for slow tailcalls is
+                    // VSD arg.
+                    assert(vsdArg == nullptr);
+                    vsdArg = arg->use;
+#ifndef DEBUG
+                    break;
+#endif
+                }
+            }
+
+            assert(vsdArg != nullptr);
+            GenTreeCall::Use** ptr = &call->gtCallArgs;
+            while ((*ptr) != vsdArg)
```

Comment:
Seems like you could avoid the loop by keeping track of the arg index instead of the use pointer.

Reply:
Is there a way to index the args?

Reply:
Ah, right. I think that the args in `fgArgInfo` are not supposed to be sorted yet so the index should match the real arg index. But I'm not sure if that is 100% guaranteed so it's probably best to leave it as is. Call morphing's a mess anyway.